### ISSUE TITLE
Liquidsoap 1.3.1 stops track randomly

### ISSUE THREAD

**Issue by xabispacebiker** [LibreTime/libretime#312]
Following my liquidsoap 1.3.1 debugging I found that liquidsoap stops tracks randomly. 

Checking on the logs I see that just at the moment when liquidsoap silences the stream, the current file has been copied over. 

As shown in the logs the track (with id 16252) is created at 21:32, then it is copied again for the second time at 21:39:59. Just at that time, the track is playing, so liquidsoap streams silence until the next track.

`tail -f /var/log/airtime/pypo/pypo.log | grep "16252"`

```
2017-09-28 21:32:00,309 [pypofile] [INFO ]  copying from http://host:port/rest/media/16252 to local cache /var/tmp/airtime/pypo/cache/scheduler/16252.mp3
2017-09-28 21:32:00,310 [pypofile] [INFO ]  http://host:port/rest/media/16252/download
2017-09-28 21:39:59,213 [pypofile] [INFO ]  copying from http://host:portrest/media/16252 to local cache /var/tmp/airtime/pypo/cache/scheduler/16252.mp3
2017-09-28 21:39:59,216 [pypofile] [INFO ]  http://host:port/rest/media/16252/download
```

It's weird because in the file `/usr/local/lib/python2.7/dist-packages/airtime_playout-1.0-py2.7.egg/pypo/pypofile.py` there is a check not to copy the track if it already exists unless it is a different size. The logs do not registry any 16252 file deletion, so I assume the file is copied over the original, because it has a different size, forcing liquidsoap to silence the stream.

**Comment by hairmare** [comment]
I think I've had this issue on occasion as well, in most cases it cleared when I started to use more tracks to figure out if the issue was specific to a track. My bet is this also affects older liquidsoap versions.

A safe way to work around it might be to only copy the file if it isn't already available. Is there a reason the new file in your case suddenly has a different size? Shouldn't a file with the same ID have the same size since since it's the same file?

**Comment by xabispacebiker** [comment]
It's happening too often, very annoying.
Maybe for some reason python is not returning the right sizes?.... 
By now I have removed the size check, just for testing. I will report what happens.

**Comment by hairmare** [comment]
At a glance it looks like getting rid of the check is the right solution. I don't see a reason we need to check the file sizes matching as the main reason for when to keep the cache. 
https://github.com/LibreTime/libretime/blob/b5b24f34641490599851c7b5039bfc25cccd94ca/python_apps/pypo/pypo/pypofile.py#L49-L56
Another possible improvement might also be to download the file to a tmpdir and move it instead of directly overwriting the cached file from the main http client response loop.
https://github.com/LibreTime/libretime/blob/b5b24f34641490599851c7b5039bfc25cccd94ca/python_apps/pypo/pypo/pypofile.py#L73-L84
Anyway, I'm looking forward to your results.

**Comment by xabispacebiker** [comment]
I got it. The problem is that the size of the source file is taken from the database (probably calculated the first time by silan on upload), and the size of the destination file is calculated through `os.path.getsize` function. Once it is calculated, if the sizes on source and destination differ, pypo tries to write back the new value to the database, so next time it won't be overriden again.

After cleaning some code in pypofile.py I started getting the following error which pointed me on the right direction:
`2017-09-30 18:54:41,338 [pypofile] [ERROR]  Could not update media file 11408 with file size and md5 hash`
(I am wondering why I didn't see that error before, if this is the culprit of the bug)

The error above means that pypofile realized the sizes differ, and attempted to write it back to database, failing, in consecuence, pypo tried to write the same file over and over again  until it was removed from queue.

Shit .. This bug caused a lot of work for both, the processor and the disk !!!

I noticed that silence was happening too on files that didn't got silenced when they were played before. The reason is that the file had to be playing at the time of pypo overwriting it. Liquidsoap stops playing it and keeps waiting for the next song.

Right now I am doing some test to find the best solution
